types/document.d.ts

```diff
@@ -216,12 +222,12 @@ declare module 'mongoose' {
     set(value: any): this;
 
     /** The return value of this method is used in calls to JSON.stringify(doc). */
-    toJSON(options: ToObjectOptions & { flattenMaps: false }): LeanDocument<this>;
-    toJSON(options?: ToObjectOptions): FlattenMaps<LeanDocument<this>>;
+    toJSON(options: ToObjectOptions & { flattenMaps: false }): LeanDocument<DocType>;
```

Comment:
Good change :+1:

Reply:
Well, I thought it was, but it turns out it breaks a few cases -- specifically cases where someone `extends Document`.

I'm still working through the tests

Reply:
After some tinkering I've made the following changes to `toJSON` and `toObject`:

* I removed the redundant signatures -- the generic version has a default template type, so there is no need to specify a type without it, that just confuses things and could throw off type inference in some cases
* I have left it using LeanDocument<DocType>, though that can be overridden with a template.

The downside to this is  that if your interface extends Document instead of using the new recommended way then you'll get a `any` type out of both methods unless you manually provide a type to the generic.

On the positive side, though, I could not find any way to use `this` in the return type which did not cause circular reference issues with my code in multiple places -- this change fixes that. Additionally, the "worst case" is that you get an `any` type out, so it won't break existing code it just may not catch issues that someone would expect it to catch. That can be mitigated by either migrating to the recommended method or manually providing the type.